tools/github_update.py

```diff
@@ -96,118 +105,237 @@ def _get_repo_path(self):
+        state should be one of the values in 'VALID_STATES'.'''
+        if not context_label:
+            context_label = self._default_context_label
+        if not context_label:
+            raise Exception('Either update() call must provide a context, ' +
+                            'or a default context must have been set')
+
+        start_time_path = os.path.join(
+            tempfile.gettempdir(),
+            'github_update-{}'.format(re.sub('[^0-9a-zA-Z]+', '_', context_label)))
+        if state == PENDING_STATE:
+            # store start time to tmp file:
+            start_time_file = open(start_time_path, 'w')
+            start_time_file.write('{}\n'.format(str(time.time())))
+            start_time_file.flush()
+            start_time_file.close()
+        elif os.path.isfile(start_time_path):
+            # when available, get start time from tmp file and include duration in message:
+            try:
+                start_time = float(open(start_time_path).read().strip())
+                # omit fractional seconds from duration:
+                message = '[{}] {}'.format(datetime.timedelta(seconds=round(time.time()) - start_time), message)
+                os.remove(start_time_path)
+            except:
+                message = '[?] {}'.format(message)
+
+        self._api.set_commit_status(context_label, state, message, details_url)
+
+
+def _get_details_link_url():
+    '''returns the url to be included as the details link in the status'''
+    details_url = os.environ.get('GITHUB_COMMIT_STATUS_URL', '') # custom URL via env
+    if not details_url:
+        details_url = os.environ.get('BUILD_URL', '') # provided by jenkins
+        if details_url:
+            details_url += 'console'
+    return details_url
+
+
+def _should_access_github():
+    if 'WORKSPACE' not in os.environ:
+        # not running in CI. skip actually sending anything to GitHub
+        return False
+    if os.environ.get('GITHUB_DISABLE', ''):
+        # environment has notifications disabled. skip actually sending anything to GitHub
+        return False
+    if not (os.environ.get('GITHUB_COMMIT_STATUS_URL') or os.environ.get('BUILD_URL')):
+        # CI job was not triggered by a PR
+        return False
+    return True
 
 
 def print_help(argv):
-    logger.info('Syntax: {} <state: pending|success|error|failure> <context_label> <status message>'.format(argv[0]))
+    logger.info('Syntax:')
+    logger.info('- Update: {} <state: {}> <context_label> [a status message ...]'.format(
+        argv[0], '|'.join(VALID_STATES)))
+    logger.info('- Reset pending: {} reset [a replacement message ...]'.format(argv[0]))
+
+
+def reset_states(updater, message):
+    if not _should_access_github():
+        # reset disabled due to local build, exit silently
+        return 0
+
+    contexts = [context for context in updater.list_contexts() if context not in BLACKLISTED_CONTEXT_LABELS]
+    if not contexts:
```

Comment:
The check is just to avoid the following message saying that we're resetting statuses, if there are no statuses to be reset